### ISSUE TITLE
On SetQueryParams, respect JsonProperty settings

### ISSUE THREAD

**Issue by thiagolunardi** [tmenier/Flurl#341]
I have a situation that I have the following parameter object:

```
     public class FooParams
     {
        [JsonProperty("page")]
        public int Page { get; set; }
        [JsonProperty("per_page")]
        public int PerPage { get; set; }
        [JsonProperty("term")]
        public string SearchTerm { get; set; }
     }
```

So, when setting this object as query parameter like this:

```
var params = new FooParams
{
    Page = 1,
    PerPage = 10,
    SearchTerm = "SpaceGhost"
};
var url = "http://www.mysite.com".SetQueryParams(foo);
```

I expect to have:
`http://www.mysite.com?page=1&per_page=10&term=SpaceGhost`

**Comment by tmenier** [comment]
I wouldn't expect that, because query parameters are not JSON encoded, so they never hit the JSON serializer. They're URL-encoded.

**Comment by thiagolunardi** [comment]
The values are URL encoded, for sure, but not the keys. And the same way that I can use the JSON serializer to deserialize using Flurl.Http, I need to follow the same rule to serialize.

At least this is a situation that I'm going throw now. I have a response object which is also returning the URL for an item detail. And the serialized URL must match the format that the other endpoint is expecting. 

I confess that I don't like the idea to use reflection, but I couldn't figure out a best way to achieve that in an easier way.

**Comment by tmenier** [comment]
My point was that using `[JsonProperty]` for this is a bit confusing because building a query string has nothing to do with JSON. That's why I don't think I want to add this to the core.

I'm not sure if I understand the context well enough to suggest a work-around, but how about an extension method like this?

```c#
public static Url SetSearchParams(this string url, int page, int perPage, string searchTerm) {
    return url.SetQueryParams(new { page, per_page = perPage, term = searchTerm });
}
```

**Comment by thiagolunardi** [comment]
Yes, you understood properly, and yes, that is the workaround that I'm doing. You're totally right, query parameters have nothing to do with JSON, but I'd like to find a generic way to specify that instead of using JSON. Maybe a Flurl custom attribute then?

Let's say I'm writing an SDK for a third party service API, and this API is expecting a properties snake case (eg. `per_page`). I'd like not to have to work, at c# side, with a property named as `instance.per_page`, but `instance.PerPage`, and just at turning this "DTO" object to a query param format, to parse as `per_page` instead.

**Comment by tmenier** [comment]
The SDK could do something similar to the example I provided, just as easily as what you did in your project. Flurl allows for expressing key-value pairs as objects in several areas (headers, cookies, form data, etc). If I provided this functionality for one of them, why shouldn't I provide it for all? But my intent was that object notation for these things is there for convenience, i.e. to save keystrokes, and that people would use anonymous object the overwhelming majority of the time. JSON payloads are _much_ different in that they're potentially deeply nested, complex objects with many more scenarios where you'd want to create classes and work with them on the C# side in interesting ways.

In short, I understand the the scenario, and I think it's great that you have an implementation that works for you and can easily be added to your project as an extension method. I just don't agree that this is _generally_ useful enough to add to the core.
